### PR TITLE
Data Flow Update 2023

### PR DESCRIPTION
**Pull request by RemingtonRohel**
## Summary
This pull request encapsulates several major changes to the data flow. These major changes are:

* Modifying data flow to work with Borealis v0.7
* Migrating the campus data flow from superdarn-cssdp to new sdc-serv computer
* Adding pyDARNio dmap check for all files before distributing them
* Moving all dmap file conversion to campus (files were previously converted at site)

#### Borealis computer data flow

* Modified script to work with new Borealis directory structure
* Added check to ensure SOURCE directory is defined
* Changed permissions on files stored on NAS to be `rwxrwxr--` to prevent NAS permission issues 

#### Site Linux data flow

* $SDCOPY on each site computer was changed to `dataman@sdc-serv.usask.ca`
* Changed permissions on files stored on NAS to be `rwxrwxr--` to prevent NAS permission issues 
* Edited `config.sh` file to move all site's dmap generation to campus instead of at site (i.e. running `convert_on_campus` for all sites)
* Bugfix for zlib compression error and boolean fields stemming from deepdish

#### Campus data flow

* Moved all data flow responsibility from superdarn-cssdp to sdc-serv. All data directories were on the NAS, so they just had to be mounted to sdc-serv identically to superdarn-cssdp.
* A pyDARNio dmap check is performed for all dmap files before distribution. This check is done by reading the file into pyDARNio. If the file fails the check, the dmap and corresponding rawacf array file are both moved to a conversion failure directory for manual inspection.

## Testing
Borealis and Site Linux data flow has been running at each site for the past few weeks, and appears to be working correctly.

Campus data flow has been tested via manual runs of each script.

### CODE REVIEW

**Comment by tjk584** [comment]
Moved check from `convert_on_campus` to `distribute_borealis_data`, as not all sites run `convert_on_campus`.

**Comment by tjk584** [comment]
Expanded this pull request to cover moving the data flow from SuperDARN-CSSDP to the new sdc-serv computer.

Data flow is set up to run fully on sdc-serv, and stage data for other institutions/our mirror in the same locations.

Hasn't been tested yet - waiting for external IP on our sdc-serv.

**Comment by tjk584** [comment]
This branch has been running for each site for several days now. Everything seems to be working.

File: config.sh [hunk 1]
```diff
@@ -17,22 +17,22 @@ readonly NAS_SITES=("sas" "pgr" "inv" "cly" "rkn")
...
 readonly RAWACF_SITES=("sas" "pgr" "inv" "cly" "rkn")
-readonly BFIQ_SITES=("pgr")
+readonly BFIQ_SITES=()
 readonly ANTENNAS_IQ_SITES=("sas" "pgr" "inv" "cly" "rkn")
 
 # Special case sites. If a site is specified here, the data flow will be slightly altered to 
 # accomodate the site
 readonly LOW_MEMORY_SITES=("pgr" "inv" "cly" "rkn" "sas")
-readonly LOW_BANDWIDTH_SITES=("cly" "rkn")
+readonly CONVERT_ON_CAMPUS_SITES=("sas" "pgr" "inv" "cly" "rkn")
```
Comment (RemingtonRohel) [inline config.sh:29]: Should we change the behaviour so that these fields can be removed? Seems a bit redundant now.

File: config.sh [hunk 2]
```diff
@@ -17,22 +17,22 @@ readonly NAS_SITES=("sas" "pgr" "inv" "cly" "rkn")
...
 
 ###################################################################################################
 
 # Define variables needed for telemetry
-readonly TELEMETRY_DIR="/home/telemetry/data_flow_logs"
-readonly TELEMETRY="telemetry@chapman.usask.ca"
-readonly TELEMETRY_RSH="ssh -p 2222"
+readonly TELEMETRY_DIR="/home/logman/data_flow_logs"
+readonly TELEMETRY="logman@sdc-serv.usask.ca"
```
Comment (RemingtonRohel) [inline config.sh:35]: Can we define these locally in the `.profile` on the computer that runs this script?

File: library/data_flow_functions.sh [hunk 3]
```diff
@@ -26,6 +26,88 @@ ord() {
...
+
+	array_filename=$(basename $1)
+	array_directory=$(dirname $1)
+	# Check that the filename given is a valid dmap file name
+	if [[ ! "$array_filename" =~ ^[0-9]{8}.[0-9]{4}.[0-9]{2}.[[:lower:]]{3}.[0-9]+.rawacf.hdf5$   ]]; then
+		printf "get_dmap_name(): Invalid filename - $array_filename isn't a valid array file name\n"
+		return 1
+	fi
+
+	# Move the resulting files if all was successful then remove the source site file.
```
Comment (RemingtonRohel) [inline library/data_flow_functions.sh:53]:
```suggestion
```

File: library/data_flow_functions.sh [hunk 4]
```diff
@@ -26,6 +26,88 @@ ord() {
...
+	# Check function was called correctly
+	if [[ $# -ne 1 ]]; then
+		printf "get_array_name(): Invalid number of arguments\n" 
+		return 1
+	fi
+
+	dmap_filename=$(basename $1)
+	dmap_directory=$(dirname $1)
+	# Check that the filename given is a valid dmap file name
+	if [[ ! "$dmap_filename" =~ ^[0-9]{8}.[0-9]{4}.[0-9]{2}.[[:lower:]]{3}.[[:lower:]]+.rawacf.bz2$   ]]; then
```
Comment (RemingtonRohel) [inline library/data_flow_functions.sh:89]: Can it also be configured to work if the file is not bzipped?

File: site-linux/remove_record.py [hunk 5]
```diff
@@ -40,30 +40,28 @@ def usage_msg():
 def remove_records_parser():
     parser = argparse.ArgumentParser(usage=usage_msg())
     parser.add_argument("borealis_site_file", help="Path to the array file that needs correction.")
-    parser.add_argument("--record_name", default='', help="Group name in the hdf5 file to remove (Borealis record name).")
+    # parser.add_argument("--record_name", default='', help="Group name in the hdf5 file to remove (Borealis record name).")
```
Comment (RemingtonRohel) [inline site-linux/remove_record.py:43]: Remove?

File: site-linux/remove_record.py [hunk 6]
```diff
@@ -40,30 +40,28 @@ def usage_msg():
...
 if __name__ == '__main__':
     parser = remove_records_parser()
     args = parser.parse_args()
-    if args.record_name:
-        remove_record(args.borealis_site_file, args.record_name)
-    else:
-        find_borealis_sequence_errors(args.borealis_site_file)
+    # if args.record_name:
+    #     remove_record(args.borealis_site_file, args.record_name)
+    # else:
```
Comment (RemingtonRohel) [inline site-linux/remove_record.py:66]: Remove?

File: campus/distribute_borealis_data [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    file_name=$(basename $file)
+
+    # TODO: Review changing of groups and permissions differring across destinations
+
+    # Flag will be > 0 if any transfers fail since verify_transfer returns 1 for failed transfer
+    transfer_flag=0
+
+    # Place file in vtsd outgoing
+    chgrp --verbose $DATA_GROUP $file
+
```
Comment (RemingtonRohel) [inline campus/distribute_borealis_data:153]:
```suggestion
```

File: campus/distribute_borealis_data [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    cp --preserve --verbose $file $VT_STAGING_DIR
+    verify_transfer $file "${VT_STAGING_DIR}/${file_name}" 
+    return_value=$?
+    transfer_flag=$(($transfer_flag + $return_value))
+
+    # Place file in BAS outgoing
+    chgrp --verbose $DATA_GROUP $file
+
```
Comment (RemingtonRohel) [inline campus/distribute_borealis_data:161]:
```suggestion
```

Comment (RemingtonRohel) [inline campus/distribute_borealis_data:161]:
```suggestion
```
Redundant due to line 152?

**Review by RemingtonRohel — COMMENTED** [review]
Some small changes but overall looks very nice.

File: config.sh (hunk 2 above)
Comment (tjk584) [reply config.sh:35]: The sdc-serv IP is already defined in .profile on all computers, so I can just modify this to get the IP from .profile

File: library/data_flow_functions.sh (hunk 4 above)
Comment (tjk584) [reply library/data_flow_functions.sh:89]: Yup, I've modified it to work with or without the `.bz2` suffix

File: config.sh (hunk 1 above)
Comment (tjk584) [reply config.sh:29]: We may want to keep the option to convert dmap files at site. Either way, this should be a separate issue/PR

**Comment by tjk584** [comment]
Running changes on PGR to test.

**Comment by tjk584** [comment]
Data flow is working on PGR, good to merge to main now.

**Pull request merged**